plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,152 @@
+
+    public static final String COMMITTED_MARKER_FILE_NAME = "committed";
+    public static final String DATA_FILE_SUFFIX = ".data";
+
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(FileSystemExchangeSink.class).instanceSize();
+    private static final int INTEGER_INSTANCE_SIZE = ClassLayout.parseClass(Integer.class).instanceSize();
+
+    private final ExchangeStorage exchangeStorage;
+    private final URI outputDirectory;
+    private final int outputPartitionCount;
+    private final Optional<SecretKey> secretKey;
+
+    @GuardedBy("this")
+    private final Map<Integer, SliceOutput> outputs = new HashMap<>();
+    @GuardedBy("this")
+    private boolean committed;
+    @GuardedBy("this")
+    private boolean closed;
+
+    public FileSystemExchangeSink(ExchangeStorage exchangeStorage, URI outputDirectory, int outputPartitionCount, Optional<SecretKey> secretKey)
+    {
+        this.exchangeStorage = requireNonNull(exchangeStorage, "exchangeStorage is null");
+        this.outputDirectory = requireNonNull(outputDirectory, "outputDirectory is null");
+        this.outputPartitionCount = outputPartitionCount;
+        this.secretKey = requireNonNull(secretKey, "secretKey is null");
+    }
+
+    @Override
+    public synchronized void add(int partitionId, Slice data)
+    {
+        checkArgument(partitionId < outputPartitionCount, "partition id is expected to be less than %s: %s", outputPartitionCount, partitionId);
+        checkState(!committed, "already committed");
+        if (closed) {
+            return;
+        }
+        SliceOutput output = outputs.computeIfAbsent(partitionId, this::createOutput);
+        output.writeInt(data.length());
+        output.writeBytes(data);
+    }
+
+    private SliceOutput createOutput(int partitionId)
+    {
+        URI outputPath = outputDirectory.resolve(partitionId + DATA_FILE_SUFFIX);
+        try {
+            return exchangeStorage.createSliceOutput(outputPath, secretKey);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return INSTANCE_SIZE +
+                estimatedSizeOf(outputs, (ignored) -> INTEGER_INSTANCE_SIZE, SliceOutput::getRetainedSize);
+    }
+
+    @Override
+    public void finish()
+    {
+        if (closed) {
+            return;
+        }
+        try {
+            for (SliceOutput output : outputs.values()) {
+                try {
+                    output.close();
+                }
+                catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
+            try {
+                exchangeStorage.createFile(outputDirectory.resolve(COMMITTED_MARKER_FILE_NAME));
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
```

Comment:
why the internal `try`/`catches`. Those do not buy you anything it looks. You are still catching everything above.

Reply:
I think it's mainly to make sure we call abort in case of failures?